fix: handle awaitTermination result and ensure proper ExecutorService shutdown

2897robo opened the pull request:
## What does this PR do?

This pull request addresses a high severity issue (java:S2095) reported by SonarCloud in the `App.java` file of the `leader-followers` module.

The issue was:

> **"Use try-with-resources or close this 'ExecutorService' in a 'finally' clause."**  
> [(See SonarCloud rule java:S2095)](https://rules.sonarsource.com/java/RSPEC-2095)

## What was the problem?

The `ExecutorService` created using `Executors.newFixedThreadPool(...)` was **not guaranteed to shut down properly**, which could lead to:
- Resource leakage
- Threads not terminating gracefully
- Reliability issues in long-running systems

## What does this PR change?

- Wrapped `ExecutorService` usage inside a `try-finally` block
- Handled the return value of `awaitTermination(...)` to check whether the shutdown was completed within the timeout
- Added a fallback `shutdownNow()` in the `finally` block
- This fix is **in line with the best practices** for managing ExecutorService lifecycle in Java

## Why not try-with-resources?

`ExecutorService` does **not implement `AutoCloseable`**, so it **cannot be used** with `try-with-resources`.  
Hence, we use `try-finally` as the proper alternative here.

## References

- Rule: [java:S2095 - Resources should be closed](https://rules.sonarsource.com/java/RSPEC-2095)
- SonarCloud issue: [AZXOQpt67L73unJ7u3kh](https://sonarcloud.io/project/issues?impactSeverities=HIGH&issueStatuses=OPEN%2CCONFIRMED&id=iluwatar_java-design-patterns&open=AZXOQpt67L73unJ7u3kh)

---

✅ **Fixes**: #2865  
🧹 **Impact**: Improves resource handling, stability, and SonarCloud quality gates

2897robo commented:
Hi @iluwatar, kindly reminding for review when you get a chance 🙂  
This PR fixes a SonarCloud high severity issue and passed all checks.  
Thanks in advance!

iluwatar commented on leader-followers/src/main/java/com/iluwatar/leaderfollowers/App.java:
```diff
@@ -68,11 +68,18 @@ public static void main(String[] args) throws InterruptedException {
+      addTasks(taskSet);
+      boolean terminated = exec.awaitTermination(2, TimeUnit.SECONDS);
+      if (!terminated) {
+        System.out.println("Executor did not terminate in the given time.");
```
Use Lombok's @Slf4j annotation for logging

2897robo commented:
Hi @iluwatar, thank you for the feedback!

✅ I've addressed your review comment by defining the SLF4J logger explicitly instead of using Lombok's `@Slf4j`.
✅ Also confirmed the proper use of `try-finally` since `ExecutorService` is not `AutoCloseable`.

Let me know if there's anything else I should improve 🙌

iluwatar commented on leader-followers/pom.xml:
```diff
@@ -47,6 +47,12 @@
       <artifactId>junit-jupiter-engine</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.projectlombok</groupId>
+      <artifactId>lombok</artifactId>
+      <version>1.18.24</version>
+      <scope>provided</scope>
+    </dependency>
```
The dependency already exists in parent pom.xml, so no need to declare it here

iluwatar commented on leader-followers/src/main/java/com/iluwatar/leaderfollowers/App.java:
```diff
@@ -1,61 +1,16 @@
- */
+@Slf4j
 public class App {
+  private static final Logger log = LoggerFactory.getLogger(App.class);
```
No need to declare the logger like this. After inserting the annotation you can just log like `LOGGER.info("hello")`

2897robo commented:
✅ Updated code based on review:
- Removed duplicate Lombok dependency (exists in parent POM)
- Removed manual logger definition
- Using Lombok's @Slf4j annotation with LOGGER.warn() syntax

Let me know if there's anything else to revise.
Thanks for reviewing as always! 🙌

2897robo commented:
Hi @iluwatar! 🙌  
I’ve updated the PR based on all your feedback:

- ✅ Removed the redundant Lombok dependency (parent POM already includes it)
- ✅ Removed manual logger definition
- ✅ Replaced it with Lombok's @Slf4j and updated to LOGGER.warn(...) style

All checks have passed ✅ and there are no conflicts with base branch.  
When you have a moment, I’d really appreciate a final review or merge.  
Thanks again for your time and guidance! 🙏

2897robo commented:
Thank you so much for reviewing and merging! 🙏  
It was a valuable learning experience to contribute and receive feedback.  
I’ll continue contributing whenever I can!

Much appreciated! 🚀
